### ISSUE TITLE
About the parameter sharing and training

### ISSUE THREAD

**Issue by liweijun-haha** [neulab/guided_summarization#6]
From “forward” functions of GuidedBARTModel and GuidedTransformerEncoder, I can understand how to share parameters in terms of model structure. However, I am confused about how the gradients flow and how to update the shared and unshared parameters. According to the paper, the unshared parameters are trained separately, but in the code, it seems that the source document and guidance signal are input together and the parameters are updated once in a step?
Thanks.

**Comment by zdou0830** [comment]
Hi, all the parameters are trained jointly. Could you point me to where we mention the unshared parameters are trained separately?

**Comment by liweijun-haha** [comment]
Thanks for your reply.
I still don't understand how the gradients flow. If there is only one part of input, it is easy to understand. 
However, the input consists of two parts, the source document and guidance signal. How to jointly update the encoding layers?

**Comment by zdou0830** [comment]
你可以想象成有两个独立的encoder分别做back propagation，在更新的时候share的部分gradient相加，不share的部分gradient不变。如果能理解recurrent neural network怎么运作的话应该这里也可以理解。

**Comment by yumoxu** [comment]
Hi @zdou0830, thanks for your great work! 

In the `class GuidedBARTModel` defined [here](https://github.com/neulab/guided_summarization/blob/a8956dfe5e77650aae9015bc3c569ef03d224dac/bart/fairseq/models/bart/guided_model.py#L42), it seems to me the same encoder is used to encode both source (see [here](https://github.com/neulab/guided_summarization/blob/a8956dfe5e77650aae9015bc3c569ef03d224dac/bart/fairseq/models/bart/guided_model.py#L139-L144)) and guidance (see [here](https://github.com/neulab/guided_summarization/blob/a8956dfe5e77650aae9015bc3c569ef03d224dac/bart/fairseq/models/bart/guided_model.py#L162-L167)). 

There are several lines containing `self.f1`and `self.f2`, which I assume are two separate encoding layers, but these lines were commented out (see [here](https://github.com/neulab/guided_summarization/blob/a8956dfe5e77650aae9015bc3c569ef03d224dac/bart/fairseq/models/bart/guided_model.py#L168-L170)).

Does this mean in this Bart model, the same encoder (initialized with pre-trained parameters) is shared between source and guidance encoding, and there is no separate parameters/training? Please let me know if there is anything I missed.

Thank you!

**Comment by zdou0830** [comment]
Hi @yumoxu, thanks for the question!

Sorry for the messy code, if you look at https://github.com/neulab/guided_summarization/blob/a8956dfe5e77650aae9015bc3c569ef03d224dac/bart/fairseq/models/guided_transformer.py#L435-L438, I used layer 1-12 to encode the source document and layer 1-11, 13 to encode the guidance signal, so layer 12 and 13 are separate.

**Comment by yumoxu** [comment]
@zdou0830 Smart implementation! Thank you for clarifying!

**Comment by liweijun-haha** [comment]
@zdou0830 明白了，感谢您的解答！
